File: src/leetcode/m16/ThreeSumClosest.java
/*
 * Copyright (c) 2024 THGEEK.COM. All rights reserved.
 */

package leetcode.m16;

import java.util.Arrays;

/**
 * LeetCode 16
 */
public class ThreeSumClosest {
    public int threeSumClosest(int[] nums, int target) {
        Arrays.sort(nums); // Sort the array
        int closestSum = Integer.MAX_VALUE / 2; // Initialize to a large value

        for (int i = 0; i < nums.length - 2; i++) {
            int left = i + 1, right = nums.length - 1;

            while (left < right) {
                int currentSum = nums[i] + nums[left] + nums[right];
                if (currentSum == target) {
                    return currentSum;
                }

                // Update the closest sum if the current sum is closer to the target
                if (Math.abs(currentSum - target) < Math.abs(closestSum - target)) {
                    closestSum = currentSum;
                }

                // Adjust pointers based on comparison with target
                if (currentSum < target) {
                    left++;
                } else {
                    right--;
                }
            }
        }

        return closestSum;
    }
}

File: src/leetcode/m16/ThreeSumClosest.go
package m16

import (
	"math"
	"sort"
)

// LeetCode 16
func threeSumClosest(nums []int, target int) int {
	sort.Ints(nums)
	arrLen := len(nums)

	closestSum := math.MaxInt32
	for i := 0; i < arrLen-2; i++ {
		left, right := i+1, arrLen-1
		for left < right {
			sum := nums[i] + nums[left] + nums[right]
			if sum == target {
				return sum
			} else if sum > target {
				right--
			} else {
				left++
			}
			if abs(sum-target) < abs(closestSum-target) {
				closestSum = sum
			}
		}
	}
	return closestSum
}

func abs(a int) int {
	if a < 0 {
		return -a
	}
	return a
}
